skbio/tree/_gme.py

```diff
@@ -0,0 +1,452 @@
+
+    >>> from skbio.tree import TreeNode
+
+    Perform greedy minimum evolution (GME).
+
+    >>> tree = gme(dm)
+    >>> print(tree.ascii_art())
+                        /-monkey
+    -human--- /--------|
+                       |          /-pig
+                        \--------|
+                                 |          /-chicken
+                                  \--------|
+                                            \-rat
+
+    Notice that, unlike neighbor joining, the tree is rooted at a taxa/leaf node.
+    This will allow it to have nearest neighbor interchange performed on it without
+    needing to re-root the tree.
+
+    """
+    if dm.shape[0] < 3:
+        raise ValueError(
+            "Distance matrix must be at least 3x3 to "
+            "generate a minimum evolution tree."
+        )
+
+    # count the number of taxa
+    n = len(dm.ids)
+
+    # create initial nodes
+    root_name, node_a_name, node_b_name = dm.ids[0], dm.ids[1], dm.ids[2]
+    root = TreeNode.read(["()%s;" % (root_name)])
+    node_a = TreeNode.read(["%s;" % (node_a_name)])
+    node_b = TreeNode.read(["%s;" % (node_b_name)])
+    root_child = root.children[0]
+    root_child.append(node_a)
+    root_child.append(node_b)
+
+    # loop over rest of nodes for k = 4 to n
+    for k in range(3, n):
+        taxa = root.count(tips=True) + 1
+        # calculate average distance matrix between subtrees of T_(k-1)
+        adm = _average_distance_matrix(root, dm)
+        # create node for taxa k
+        node_k_name = dm.ids[k]
+        connecting_node = TreeNode.read(["(%s);" % (node_k_name)])
+        node_k = connecting_node.find(node_k_name)
+        # create average distance lists for subtrees of T_(k-1)
+        ordered = list(root.postorder(include_self=False))
+        lowerlist = _lower_subtree_list(ordered, node_k, dm)
+        upperlist = _upper_subtree_list(ordered, node_k, taxa, dm)
+        # Initialize the list of edge parent nodes for computation.
+        # The tuple is a pair of a TreeNode object and the value of the tree
+        # length for attaching at the edge consisting of the node and its parent.
+        # Here, we start with setting the initial length value of the edge
+        # defined by the root and its unique descendant to 0.
+        edge_list = [(root_child, 0)]
+        for a, i in edge_list:
+            if a.is_tip():
+                continue
+            a1, a2 = a.children
+            value1 = _edge_attachment_length(
+                a1, lowerlist, upperlist, ordered, taxa, adm
+            )
+            value2 = _edge_attachment_length(
+                a2, lowerlist, upperlist, ordered, taxa, adm
+            )
+            e1 = (a1, value1 + i)
+            e2 = (a2, value2 + i)
+            edge_list.append(e1)
+            edge_list.append(e2)
+        edge_nodes, edge_values = list(zip(*edge_list))
+        # find the edge with minimum length after edge attachment
+        for i in range(len(edge_values)):
```

Comment:
This code may not be optimal. See here for the common methods: https://stackoverflow.com/questions/2474015/